nominal/core/event.py

```diff
@@ -0,0 +1,134 @@
+from nominal.core.dataset import Dataset
+from nominal.core.log import LogSet
+from nominal.core.video import Video
+from nominal.ts import IntegralNanosecondsUTC, _SecondsNanos
+
+
+@dataclass(frozen=True)
+class Event(HasRid):
+    uuid: str
+    asset_rids: Sequence[str]
+    name: str
+    start: IntegralNanosecondsUTC
+    duration: timedelta
+    type: EventType
+
+    _clients: _Clients = field(repr=False)
+
+    class _Clients(
+        Attachment._Clients,
+        Asset._Clients,
+        Connection._Clients,
+        Dataset._Clients,
+        LogSet._Clients,
+        Video._Clients,
+        HasAuthHeader,
+        Protocol,
+    ):
+        @property
+        def event(self) -> event.EventService: ...
+
+    def update(
+        self,
+        *,
+        name: str | None = None,
+        asset_rids: Sequence[str] | None = None,
+        start: datetime | IntegralNanosecondsUTC | None = None,
+        duration: timedelta | None = None,
+        properties: Mapping[str, str] | None = None,
+        labels: Sequence[str] | None = None,
+        type: EventType | None,
+    ) -> Self:
+        """Replace event metadata.
+        Updates the current instance, and returns it.
+        Only the metadata passed in will be replaced, the rest will remain untouched.
+
+        Note: This replaces the metadata rather than appending it. To append to labels or properties, merge them before
+        calling this method. E.g.:
+
+            new_labels = ["new-label-a", "new-label-b"]
+            for old_label in event.labels:
+                new_labels.append(old_label)
+            run = event.update(labels=new_labels)
+        """
+        request = event.UpdateEvent(
+            uuid=self.uuid,
+            asset_rids=None if asset_rids is None else list(asset_rids),
+            duration=None if duration is None else _to_api_duration(duration),
+            labels=None if labels is None else list(labels),
+            name=name,
+            properties=None if properties is None else dict(properties),
+            timestamp=None if start is None else _SecondsNanos.from_flexible(start).to_api(),
+            type=None if type is None else type._to_api_event_type(),
+        )
+        response = self._clients.event.update_event(self._clients.auth_header, request)
+        e = self.__class__._from_conjure(self._clients, response)
+        update_dataclass(self, e, fields=self.__dataclass_fields__)
+        return self
+
+    @classmethod
+    def _from_conjure(cls, clients: _Clients, event: event.Event) -> Self:
+        return cls(
+            uuid=event.uuid,
+            asset_rids=tuple(event.asset_rids),
+            name=event.name,
+            start=_SecondsNanos.from_api(event.timestamp).to_nanoseconds(),
+            duration=timedelta(
+                seconds=event.duration.seconds,
+                microseconds=event.timestamp.nanos / 1e3
+                + (0 if event.timestamp.picos is None else event.timestamp.picos / 1e6),
```

Comment:
the nanos and picos precision would be lost here, it's immediately truncated to integral precision:

```py
>>> timedelta(microseconds=5.5)
datetime.timedelta(microseconds=6)
```

Elsewhere in the library we use `IntegralNanosecondsUTC` for API responses for an unamiguously typed value (because similarly `datetime` only supports microsecond precision), maybe we should do similar here with an `IntegralNanosecondsDuration` or something? We don't meaningfully support picos yet in the client lib, so I think we could utilize something like this for now and later would need to make the change across the whole lib to utilize picos.

(so for now maybe just throw or warn if picos is not None?)

Reply:
fwiw it would probably be fine to use `timedelta | IntegralNanosecondsDuration` or similar for _inputs_, but when retrieving from the API I don't want to truncate the nanos